Translated Keeping components pure (#430)
reactjs/it.react.dev: src/content/learn/keeping-components-pure.md

---
title: Mantenere i Componenti Puri
---

<Intro>

Alcune funzioni JavaScript sono *pure.* Le funzioni pure eseguono solo un calcolo e nient'altro. Scrivendo rigorosamente i tuoi componenti solo come funzioni pure, puoi evitare un'intera classe di bug confusi e comportamenti imprevedibili man mano che la tua base di codice cresce. Per ottenere questi vantaggi, tuttavia, ci sono alcune regole che devi seguire.

</Intro>

<YouWillLearn>

* Che cosa è la purezza e come ti aiuta a evitare i bug
* Come mantenere i componenti puri, mantenendo i cambiamenti fuori dalla fase di renderizzazione
* Come utilizzare la Strict Mode per trovare errori nei tuoi componenti

</YouWillLearn>

## Purezza: Componenti Come Formule {/*purity-components-as-formulas*/}

In informatica (e in particolare nel mondo della programmazione funzionale),  [una funzione pura](https://wikipedia.org/wiki/Pure_function) è una funzione con le seguenti caratteristiche:

* **Si cura solo dei suoi affari.** Non modifica oggetti o variabili che sono esistiti prima che fosse chiamata.

> * **Same inputs, same output.** Dati gli stessi input, una funzione pura dovrebbe sempre restituire lo stesso risultato.

Reviewer: Non so se qui hai voluto mantenere il detto oppure se è una piccola dimenticanza 😃 
_"Same inputs, same output"_ - forse meglio tradurlo in _"Stessi input, stesso output"_? Cosa ne pensi?

Potresti avere già familiarità con un esempio di funzioni pure: le formule in matematica.

Considera questa formula matematica:  <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math>.

Se <Math><MathI>x</MathI> = 2</Math> allora <Math><MathI>y</MathI> = 4</Math>. Sempre. 

Se <Math><MathI>x</MathI> = 3</Math> allora <Math><MathI>y</MathI> = 6</Math>. Sempre. 

Se <Math><MathI>x</MathI> = 3</Math>, <MathI>y</MathI> non sarà a volte <Math>9</Math> o <Math>–1</Math> o <Math>2.5</Math> a seconda dell'ora del giorno o dello stato del mercato azionario.

> Se <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math> e <Math><MathI>x</MathI> = 3</Math>, <MathI>y</MathI> sarà sempre <Math>6</Math>. 

Reviewer 2: ```suggestion
Se <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math> e <Math><MathI>x</MathI> = 3</Math>, <MathI>y</MathI> sarà _sempre_ <Math>6</Math>. 
```

</Sandpack>

> Quando passi `drinkers={2}` a `Ricetta`, restituirà JSX contenente `2 tazze di acqua`. Sempre. 

Reviewer: ```suggestion
Quando passi `drinkers={2}` a `Ricetta`, questo restituirà JSX contenente `2 tazze di acqua`. Sempre. 
```

> Se si passa `drinkers={4}`, restituirà JSX contenente `4 tazze di acqua`. Sempre.

Reviewer: ```suggestion
Se si passa `drinkers={4}`, questo restituirà JSX contenente `4 tazze di acqua`. Sempre.
```

Proprio come una formula matematica.

Si potrebbe pensare ai propri componenti come ricette: se le si segue e non si introducono nuovi ingredienti durante il processo di cottura, si otterrà lo stesso piatto ogni volta. Quel "piatto" è il JSX che il componente fornisce a React per il [render.](/learn/render-and-commit)

<Illustration src="/images/docs/illustrations/i_puritea-recipe.png" alt="A tea recipe for x people: take x cups of water, add x spoons of tea and 0.5x spoons of spices, and 0.5x cups of milk" />

> ## Side Effects: Conseguenze Inaspettate {/*side-effects-unintended-consequences*/}

Reviewer: ```suggestion
## Side Effects: Conseguenze (In)aspettate {/*side-effects-unintended-consequences*/}
```

> ## Side Effects: Conseguenze (In)aspettate {/*side-effects-unintended-consequences*/}

Reviewer 2: ```suggestion
## Side Effects: conseguenze (in)aspettate {/*side-effects-unintended-consequences*/}
```
Teniamo il casing come l'originale :)

Il processo di rendering di React deve sempre essere puro. I componenti dovrebbero solo *restituire* il loro JSX e non *modificare* oggetti o variabili che esistevano prima del rendering, in quanto questo li renderebbe impuri!

Ecco un componente che viola questa regola:

<Sandpack>

```js
let guest = 0;

function Cup() {

>   // Non corretto: modificare una variabile esistente!

Reviewer: ```suggestion
  // Non corretto: sta modificando una variabile esistente!
```

</Sandpack>

Questo componente legge e scrive una variabile `guest` dichiarata esternamente. Ciò significa che **chiamare questo componente più volte produrrà JSX diversi!** E inoltre, se altri componenti leggono `guest`, produrranno anche loro JSX diversi, a seconda del momento in cui sono stati eseguiti! Ciò non è prevedibile.

Tornando alla nostra formula <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math>, anche se <Math><MathI>x</MathI> = 2</Math>, non possiamo essere certi che <Math><MathI>y</MathI> = 4</Math>. I nostri test potrebbero fallire, gli utenti sarebbero confusi, gli aerei cadrebbero dal cielo - si può vedere come ciò porterebbe a bug confusi!

> È possibile risolvere questo componente [passando "guest" come una prop:](/learn/passing-props-to-a-component):

Reviewer 2: ```suggestion
È possibile risolvere questo componente [passando `guest` come una prop:](/learn/passing-props-to-a-component):
```

Nota come l'esempio originale abbia visualizzato "Guest #2", "Guest #4" e "Guest #6" invece di "Guest #1", "Guest #2" e "Guest #3". La funzione originale era impura, quindi chiamarla due volte l'ha rotta. Ma la versione pura funziona correttamente anche se la funzione viene chiamata due volte ogni volta. **Le funzioni pure calcolano solo, quindi chiamarle due volte non cambierà nulla**--proprio come chiamare `double(2)` due volte non cambia ciò che viene restituito e risolvere  <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math> due volte non cambia il valore di <MathI>y</MathI>. Gli stessi input, gli stessi output. Sempre.

Strict Mode non ha alcun effetto in produzione, quindi non rallenterà l'app per i tuoi utenti. Per abilitare Strict Mode, puoi avvolgere il tuo componente principale all'interno di `<React.StrictMode>`. Alcuni framework lo fanno di default.

</DeepDive>

### Mutazione Locale: il Piccolo Segreto del tuo Componente {/*local-mutation-your-components-little-secret*/}

> Nell'esempio precedente, il problema era che il componente ha cambiato una variabile *preesistente* durante il rendering. Questo è spesso chiamato una **"mutazione"** per renderlo un po' più spaventoso. Le funzioni pure non mutano variabili al di fuori della portata della funzione o oggetti creati prima della chiamata - questo le rende incontaminate!

Reviewer 2: ```suggestion
Nell'esempio precedente, il problema era che il componente ha cambiato una variabile *preesistente* durante il rendering. Questo è spesso chiamato una **"mutazione"** per renderlo un po' più spaventoso. Le funzioni pure non mutano variabili al di fuori della portata della funzione o oggetti creati prima della chiamata - questo le rende impure!
```
Di solito si parla di funzioni _pure_ e _impure_

</Sandpack>

> Se la variabile `cups` o l'array `[]` fossero stati creati fuori dalla funzione `TeaGathering`, questo sarebbe stato un grande problema! Staresti cambiando un oggetto *preesistente* inserendo elementi in quell'array.

Reviewer: ```suggestion
Se la variabile `cups` o l'array `[]` fossero stati creati fuori dalla funzione `TeaGathering`, questo sarebbe stato un grande problema! Staresti cambiando un oggetto *pre-esistente* inserendo elementi in quell'array.
```

Tuttavia, è permesso perché li hai creati *durante lo stesso render*, all'interno di `TeaGathering`. Nessun codice al di fuori di `TeaGathering` saprà mai che questo è accaduto. Questo si chiama **"mutazione locale"**—è come il piccolo segreto del tuo componente.

> ## Dove _Puoi_ Causare Side Effects {/*where-you-_can_-cause-side-effects*/}

Reviewer 2: ```suggestion
## Dove _puoi_ causare side effects {/*where-you-_can_-cause-side-effects*/}
```
Lasciamo il casing come l'originale :)

Mentre la programmazione funzionale si basa pesantemente sulla purezza, ad un certo punto, in qualche posto, _qualcosa_ deve cambiare. Questo è il punto della programmazione! Questi cambiamenti, come l'aggiornamento dello schermo, l'avvio di un'animazione, la modifica dei dati, sono chiamati **side effects.** Sono cose che accadono _"al margine"_, non durante il rendering.

> In React, **side effect appartengono solitamente agli [event handlers.](/learn/responding-to-events)** Gli event handler sono funzioni che React esegue quando si esegue un'azione, ad esempio quando si fa clic su un pulsante. Anche se gli event handler sono definiti *all'interno* del componente, non vengono eseguiti *durante* il rendering! **Quindi gli event handler non devono essere puri**

Reviewer: Nelle mie translations per la pagina _"Responding to Events"_ "event handlers" é stato tradotto - resta da capire se verrà mantenuta la traduzione in italiano o se invece si potrà mantenere il termine in Inglese (come preferirei - vedo però che la tendenza è quella di tradurre tutto).

Nel caso il termine dovesse venire tradotto in Italiano, qui ci sono le mie suggestions 😃 

```suggestion
In React, i **side effects appartengono solitamente ai [gestori di eventi.](/learn/responding-to-events)** I gestori di eventi sono funzioni che React esegue quando si esegue un'azione, ad esempio quando si fa clic su un pulsante. Anche se i gestori di funzione sono definiti *all'interno* del componente, non vengono eseguiti *durante* il rendering! **Quindi i gestori di funzione non devono essere puri**
```

Reviewer 2: Sì, sarebbe meglio non tradurlo come da glossario/issue @Ago95Dev @Th3Wall 
cc @deblasis

Se hai esaurito tutte le altre opzioni e non riesci a trovare l'event handler giusto per il tuo side effect, puoi comunque allegarlo al tuo JSX restituito con una chiamata [`useEffect`](/reference/react/useEffect) nel tuo componente. Ciò indica a React di eseguirlo in seguito, dopo il rendering, quando sono consentiti side effect. **Tuttavia, questo approccio dovrebbe essere l'ultima risorsa.**

Quando possibile, cerca di esprimere la tua logica solo con il rendering. Rimarrai sorpreso di quanto lontano questo possa portarti!

<DeepDive>

> #### Perché a React Importa la Purezza? {/*why-does-react-care-about-purity*/}

Reviewer 2: ```suggestion
#### Perché a React importa la purezza? {/*why-does-react-care-about-purity*/}
```
Anche qui come l'originale :)

</DeepDive>

<Recap>

* Un componente deve essere puro, cioè: 
  * **Si cura solo dei suoi affari.** Non deve cambiare alcun oggetto o variabile esistente prima del rendering.
  * **Same inputs, same output.** Dati gli stessi input, una funzione pura dovrebbe sempre restituire lo stesso JSX. 
* Il rendering può accadere in qualsiasi momento, quindi i componenti non dovrebbero dipendere dalla sequenza di rendering l'uno dell'altro.
* Non devi mutare alcuno degli input che i tuoi componenti utilizzano per il rendering. Questo include props, stato e contesto. Per aggiornare lo schermo,  [imposta ("set") lo state](/learn/state-a-components-memory)  invece di modificare oggetti pre-esistenti.
* Cerca di esprimere la logica del tuo componente all'interno del JSX che restituisci. Quando hai bisogno di "cambiare le cose", di solito vorrai farlo in un gestore di eventi. Come ultima risorsa, puoi utilizzare `useEffect`.
* Scrivere funzioni pure richiede un po' di pratica, ma sblocca il potere del paradigma di React.

</Recap>

  
<Challenges>

> #### Riparare un Orologio Rotto {/*fix-a-broken-clock*/}

Reviewer 2: ```suggestion
#### Riparare un orologio rotto {/*fix-a-broken-clock*/}
```

</Sandpack>

In questo esempio, l'effetto collaterale (modifica del DOM) non era affatto necessario. Dovevi solo restituire JSX.

</Solution>

> #### Correggi un Profilo Rotto {/*fix-a-broken-profile*/}

Reviewer 2: ```suggestion
#### Correggi un profilo rotto {/*fix-a-broken-profile*/}
```

</Sandpack>

Ricorda che React non garantisce che le funzioni dei componenti si eseguano in un ordine particolare, quindi non puoi comunicare tra di loro impostando variabili. Tutta la comunicazione deve avvenire tramite props.

</Solution>

> #### Sistema una Cassetta Rotta {/*fix-a-broken-story-tray*/}

Reviewer 2: ```suggestion
#### Sistema una cassetta rotta {/*fix-a-broken-story-tray*/}
```

<Solution>

Nota come ogni volta che l'orologio viene aggiornato, "Crea Storia" viene aggiunto *due volte*. Ciò serve come suggerimento che abbiamo una mutazione durante il rendering--Strict Mode chiama i componenti due volte per rendere questi problemi più evidenti.

> La funzione `StoryTray` non è pura. Chiamando `push` sull'array  `stories` ricevuto (una prop!), sta mutando un oggetto che è stato creato *prima* che `StoryTray` iniziasse a renderizzare. Questo lo rende buggato e molto difficile da prevedere.

Reviewer 2: ```suggestion
La funzione `StoryTray` non è pura. Chiamando `push` sull'array `stories` ricevuto (una prop!), sta mutando un oggetto che è stato creato *prima* che `StoryTray` iniziasse a renderizzare. Questo lo rende buggato e molto difficile da prevedere.
```

</Sandpack>

> 

Reviewer 2: ```suggestion
```

</Sandpack>

> icona di pizzaGPT, una fetta di pizza

Reviewer: 🤔 è un refuso questo? 😅 
```suggestion

```

</Sandpack>

> In questo modo, la mutazione avviene localmente e la funzione di rendering rimane pura. Tuttavia, è necessario fare attenzione: ad esempio, se si provasse a modificare uno degli elementi esistenti dell'array, bisognerebbe clonare anche quelli.

Reviewer: ```suggestion
Questo mantiene la mutazione locale e la funzione di rendering rimane pura. Tuttavia, è necessario fare attenzione: ad esempio, se si provasse a modificare uno degli elementi esistenti dell'array, bisognerebbe clonare anche quelli.
```
